### ISSUE TITLE
DRAT does not identify different files with same name

### ISSUE THREAD

**Issue by lewismc** [apache/drat#24]
Say for example I have 4 files called index.html e.g.

```
lmcgibbn@LMC-032857 /usr/local/incubator-usergrid(master) $ find . -name "index.html"
./sdks/android/doc/index.html
./sdks/html5-javascript/index.html
./sdks/html5-javascript/tests/mocha/index.html
./sdks/java/doc/index.html
```

The output of DRAT is as follows

```
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/index.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/insecure-banner.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/ios.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/javascript.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/loading.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/login.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/logout.html
```

Here you can see that only one index.html has been specified! This is a bug and it should be resolved. 
A possibly mechansim for actually adding more context to the location of 'Unapproved licenses', would be for us to alter the DRAT output to something the following

```
Unapproved licenses:

  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/account.html            --> /path/to/account.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/activities.html           --> /path/to/activities
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/all-calls.html             --> etc
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/android.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/apigee_test.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/app-overview.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/appMenu.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/data.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/display-generic.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/display-roles.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/display-users.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/dogs.html
```

**Comment by lewismc** [comment]
This issue is confirmed as I can reproduce it with alternative files within the Usergrid project. Namely;

```
lmcgibbn@LMC-032857 /usr/local/incubator-usergrid(master) $ find . -name "test.html"
./sdks/html5-javascript/examples/persistence/test.html
./sdks/html5-javascript/examples/test/test.html
./sdks/html5-javascript/tests/test.html
```

where these only show up once in the DRAT output.

```
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/roles.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/shell.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/test.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/users-activities.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/users-feed.html
  /usr/local/drat/deploy/data/jobs/rat/1407484852545/input/users-graph.html
```

I am pretty positive that is a bug in RAT and not DRAT... we most likely need to file it over there.
Hows about discussion here first?

**Comment by chrismattmann** [comment]
Sounds good, thanks Lewis. One thing to realize is that DRAT partitions jobs by MIME type and into diff sub directories of size 100 (configurable as well). That said, are you sure that there actually aren't copies of e.g., index.html or test.html, but that they aren't in e.g., some other DRAT job directory for RAT? For example you can do a file manager query and see how many files come back for test.html?

**Comment by lewismc** [comment]
Based on the above I'll fire back up FileMgr and try to get to the bottom. Thanks for comments.

**Comment by chrismattmann** [comment]
@lewismc did you get a chance to check this?

**Comment by lewismc** [comment]
hi @chrismattmann no I didn't get a chance to check this out however I will update this issue once I run DRAT over the HTrace - http://incubator.apache.org/projects/htrace.html codebase. We are just making sure that the Software grant from Cloudera check's out OK, then I assume the podling will continue with setting up codebase @Apache.
I'll update once I use DRAT next.

**Comment by chrismattmann** [comment]
thanks @lewismc OK!

**Comment by lewismc** [comment]
@chrismattmann OK, I just got a chance to identify this bug.
When reviewing the recent OODT 0.8.1 RC#1 I [observed and documented](https://issues.apache.org/jira/browse/OODT-807) the following...
In my DRAT log for HTML unapproved licenses, there appeared to be only one file named async.html e.g.
```
 21 Unapproved licenses:
 22
 23   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/async.html
 24   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/images.html
 25   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/large.html
 26   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/package.html
 27   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/prerendered.html
 28   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/simple.html
```
However in fact there were two
```

```

**Comment by lewismc** [comment]
@chrismattmann just been able to verify this bug

**Comment by lewismc** [comment]
If you look at 
https://github.com/apache/oodt/blob/trunk/profile/src/main/java/org/apache/oodt/profile/handlers/lightweight/package.html
you will see that there is no license header. 
If you then look at 
https://github.com/apache/oodt/blob/trunk/xmlquery/src/main/java/org/apache/oodt/xmlquery/package.html
you will also notice no license header.
Within the DRAT report I posted for the OODT 0.8.1 RC#1 results, you will see the following results

```
 21 Unapproved licenses:
 22
 23   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/async.html
 24   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/images.html
 25   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/large.html
 26   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/package.html
 27   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/prerendered.html
 28   /usr/local/drat/deploy/data/jobs/rat/1422589552487/input/simple.html
```

This is incorrect, there should be two such entries.

**Comment by lewismc** [comment]
I don't know where this bug actually resides to be honest. I have a feeling it is possibly within RAT itself...

**Comment by lewismc** [comment]
The same is true for 

```
 23   /usr/local/drat/deploy/data/jobs/rat/1422589552402/input/build.xml
```

which actually relates to both

```
lmcgibbn@LMC-032857 /usr/local/oodt_trunk(master) $ find . -name "build.xml"

./tools/pdi_plugin/build/build.xml
./tools/pdi_plugin/build.xml
```

**Comment by chrismattmann** [comment]
so the File Manager catalog currently contains multiple files that have the same name. So that's correct. So then I checked out the MimePartitioner. What it does it then looks up all files by type, consisting of the following types:

```
   mimeTypes = ['x-java-source', 'x-c', 'javascript', 'xml', 'html', 'css', 'x-json', 'x-sh', 'x-fortran', 'csv' 'tab-separated-values', 'x-tex', 'x-asm', 'x-diff', 'x-python']
```

For each one of those types, it queries Solr, gets the info filename, file location, joins these together and sets the metadata field InputFiles to a multi-valued set for each Rat PGE task.
The RAT PGE task should write a script file in $DRAT_HOME/data/jobs/rat/<job number>/sciPgeExeScript_RatCodeAudit. An example of which is:

```
[usc-secure-wireless-062-114:jobs/rat/1425608188401] mattmann% more sciPgeExeScript_RatCodeAudit 
#!sh
export PATH=$HOME/bin/:${PATH}
shopt -s expand_aliases
alias rat="java -jar /Users/mattmann/drat/deploy/rat/lib/apache-rat-0.9.jar"
echo "Creating working dirs"
mkdir /Users/mattmann/drat/deploy/data/jobs/rat/1425608188401/input ; mkdir /Users/mattmann/drat/deploy/data/jobs/rat/1425608188401/output; mkdir /Users/mattmann/drat/deploy/data/jobs/rat/1425608188401/logs
echo "Staging input to /Users/mattmann/drat/deploy/data/jobs/rat/1425608188401/input"
cp -R `python -c "print ' '.join('/Users/mattmann/drat/src/pge/src/main/resources/bin/mime_partitioner/mime_rat_partitioner.py,/Users/mattmann/drat/src/pge/src/main/resources/bin/rat_aggregator/rat_aggregator.py,/Users/mattmann/drat/src/pge/target/classes/bin/mime_partitioner/mime_rat_partitioner.py,/Users/mattmann/drat/src/pge/target/classes/bin/rat_aggregator/rat_aggregator.py'.split(','))"` /Users/mattmann/drat/deploy/data/jobs/rat/1425608188401/input
echo "Running Apache RAT on /Users/mattmann/drat/deploy/data/jobs/rat/1425608188401/input"
rat /Users/mattmann/drat/deploy/data/jobs/rat/1425608188401/input > /Users/mattmann/drat/deploy/data/jobs/rat/1425608188401/output/rat_x-python_1425608188401.log

[usc-secure-wireless-062-114:jobs/rat/1425608188401] mattmann% 
```

_There_ is the error, @lewismc - it's on the copy step in the RAT PGE job. Looks like we need to check if the files have the same name, and if so, dedup them. Maybe just replacing '/' with '_'. Let me think.

**Comment by chrismattmann** [comment]
fixed and tested in #34
